### ISSUE TITLE
Style/StringConcatenation: flexibility settings

### ISSUE THREAD

**Issue by zverok** [rubocop-hq/rubocop#8470]
New `Style/StringConcatenation` is obviously useful, but it seems it needs more flexibility to not harm code readability.

For example, 
`"http://" + domain` → `"http://#{domain}"` 
is a good suggestion, but 
`"http://" + fetch_domain(some, params, with: COMPLEX[calculation]) + "/endpoint"` 
is more vague.

Maybe it would make sense to limit the suggestion by either concatenated expression character count, or by its complexity (number of sub-expressions or something)?..

**Comment by marcandre** [comment]
Are you suggesting to not raise offense for your second example, or to raise but not auto-correct that?

Because there's an argument for
```ruby
# bad
"http://" + fetch_domain(some, params, with: COMPLEX[calculation]) + "/endpoint"
# good
domain = fetch_domain(some, params, with: COMPLEX[calculation])
"http://#{domain}/endpoint"
```

We can't autocorrect this way, of course, but I feel an offense is still justified.
Maybe a different cop could look for `"string#{with_very_long_expressions(:inside, :a, :string)}"` and suggest to extract and name the expression...

**Comment by zverok** [comment]
> Are you suggesting to not raise offense for your second example, or to raise but not auto-correct that?

Probably "provide (some) method of limiting the offenses only to atomic (?) clauses, or to clauses of some `Max` complexity/charcter length".
I somewhat overcomplicated the example, but even something simpler, like `"http://" + domain(environment: 'production') + "/endpoint"` might seem less readable with interpolation (and not deserving one more instance var.).

Probably, "do not autocorrect when there is complex expressions, but still raise" _might_ be a sutiable compromise.

**Comment by dvandersluis** [comment]
@marcandre I'd love the 2nd example to raise but not autocorrect, because I've run into some weird autocorrections, such as

```ruby
if Rails.env.production?
  "/#{id}"
else
  "dev/#{id}"
end + '/video'
```

becoming

```ruby
"#{if Rails.env.production?
     "/#{id}"
   else
     "dev/#{id}"
end}/video"
```

Which is wrong for a number of reasons. I'd rather just fix it by extracting a variable, as you suggested.

**Comment by marcandre** [comment]
I'm 👍 in keeping offenses but adding some criteria so auto-correction does not happen.

I'm thinking these should imply no correction:
- nested interpolation
- multi-line

(PR for this welcome)

I'm personally less enthusiastic about adding config options in general / judgement about length / other complexity of code, but others may be persuaded with the right examples

**Comment by dvandersluis** [comment]
@marcandre sure! I'll try to put together a PR.
